### PR TITLE
Make Linux package universal.

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Make Linux package universal.
- Add packaging build for Linux
- Consolidate to two universal packages one for RedHat and one for deb
- Use post install and removal scripts to deal with symbolic links

NON-GOAL:  Improve libmi support.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] **This is breaking but cannot be experimental.**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7387
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: .vsts-ci/linux/templates/packaging.yml
Comment (adityapatwardhan) [inline .vsts-ci/linux/templates/packaging.yml:2]: should we just default to 16.04?

File: .vsts-ci/linux.yml
Comment (adityapatwardhan) [inline .vsts-ci/linux.yml:94]: Comment seems outdated

File: tools/packaging/packaging.psm1
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1:1288]: Remove this?

File: tools/packaging/packaging.strings.psd1
Comment (adityapatwardhan) [inline tools/packaging/packaging.strings.psd1:64]: Remove this

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: .vsts-ci/linux/templates/packaging.yml
Comment (TravisEz13) [reply .vsts-ci/linux/templates/packaging.yml:2]:
```suggestion
  pool: 'ubuntu-16.04'
```

File: tools/packaging/packaging.strings.psd1
Comment (jborean93) [inline tools/packaging/packaging.strings.psd1:14]: See https://github.com/PowerShell/PowerShell/issues/15029#issuecomment-800690344 you can't just link the `libssl` and `libcrypto` symlinks to the OpenSSL 1.1.x version and just expect it to work. It does allow the library to load but causes a hard fault in the process as soon as `libmi` tries to access any of the OpenSSL libraries in that. OpenSSL 1.1.x and 1.0.x (what `libmi` that is shipped by PowerShell is linked to) are not API and ABI compatible.

You need to either link it to OpenSSL 1.0.x and document the `compat-openssl10` (other whatever package provides OpenSSL 1.0.x on the distro) requirement for WSMan work or ship a recompiled `libmi` against OpenSSL 1.1.x so it can use a modern OpenSSL version.

File: tools/packaging/packaging.strings.psd1
Comment (jborean93) [inline tools/packaging/packaging.strings.psd1:52]: Ditto here, creating a symlink that points to the wrong OpenSSL version library is not the right thing to do here.

File: tools/packaging/packaging.strings.psd1
Comment (iSazonov) [reply tools/packaging/packaging.strings.psd1:14]: > ship a recompiled libmi against OpenSSL 1.1.x

If I remember correctly this is blocked by external team.

------------------

We could try to load libmi and the old openssl to alternative ALC. I hope AssemblyLoadContext.LoadUnmanagedDll() will do this. If yes it is not too huge work. /cc @rjmholt

Comment (jborean93) [reply tools/packaging/packaging.strings.psd1:14]: There’s a long history behind it, PowerShell can always just recompile the libraries from the same source and ship it. It’s not the team maintains it anyway as the source is missing a lot of security features and contains bugs that just aren’t focused on. But this has been commented on quite a lot of times and ultimately the team isn’t  interested in doing this. That doesn’t mean the package should point the link to the wrong OpenSSL version.

> I hope AssemblyLoadContext.LoadUnmanagedDll() will do this

I don’t think so, the linking happened at the unmanaged layer. PowerShell isn’t loading libcrypto, libssl, or even libmi directly. It’s loading libpsrpclient and whatever the native loading mechanism is loading libmi/libssl.

Ultimately you just can’t try and use libssl from 1.1.x when the calling code is expecting the symbols and API from the 1.0.x version. It might be enough to get libmi to load but as soon as you try and open a https connection it will crash the entire process. If the goal is to try and lessen the dependency on OpenSSL 1.0.x for WSMan then this is the wrong thing to do. It’s still a dependency and now you are introducing process crashes when it tried to access libssl.

Comment (iSazonov) [reply tools/packaging/packaging.strings.psd1:14]: > I don’t think so, the linking happened at the unmanaged layer. PowerShell isn’t loading libcrypto, libssl, or even libmi directly. It’s loading libpsrpclient and whatever the native loading mechanism is loading libmi/libssl.

In that case, it doesn't matter at all - it didn't work before and won't work now - nothing has changed.

Comment (jborean93) [reply tools/packaging/packaging.strings.psd1:14]: It does still matter, if we look at the PowerShell codebase we can see the various PInvoke definitions for the WSMan api are being called through [libpsrpclient](https://github.com/PowerShell/PowerShell/blob/2784add414393b0ecab798055cfbac0d363c1c1e/src/System.Management.Automation/engine/remoting/fanin/WSManNativeAPI.cs#L2339). On non-Windows this is essentially a build of https://github.com/PowerShell/psl-omi-provider and is designed to replicate the Windows WSMan layer and is dynamically linked to `libmi` which is where the actual WSMan client is implemented.

The crux of this problem is what `libmi` is dyamically linked to. In both the version provided for EL7 and EL8 we can see that they are both linked against OpenSSL 1.0.x.

```
[root@45affdb15cca /]# ldd /opt/microsoft/powershell/7/libmi.so
ldd: warning: you do not have execution permission for `/opt/microsoft/powershell/7/libmi.so'
        linux-vdso.so.1 (0x00007ffde7395000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff4980fc000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007ff497ef8000)
        libpam.so.0 => /lib64/libpam.so.0 (0x00007ff497ce8000)
        libssl.so.1.0.0 => /opt/microsoft/powershell/7/libssl.so.1.0.0 (0x00007ff497a54000)
        libcrypto.so.1.0.0 => not found
        libc.so.6 => /lib64/libc.so.6 (0x00007ff497691000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff49831c000)
        libaudit.so.1 => /lib64/libaudit.so.1 (0x00007ff497467000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007ff496f81000)
        libz.so.1 => /lib64/libz.so.1 (0x00007ff496d6a000)
        libcap-ng.so.0 => /lib64/libcap-ng.so.0 (0x00007ff496b64000)
```

_Note: `libssl` and `libcrypto` have the explict `1.0.0` extension to denote the version._

This PR (and what's currently in master) creates 2 symbolic links in the PowerShell install directory.

|Distribution|libcrypto.so.1.0.0 target|libssl.so.1.0.0 target|
|-|-|-|
|EL7|/lib64/libcrypto.so.10|/lib64/libssl.so.10|
|EL8|/lib64/libcrypto.so.1.1.1|/lib64/libcrypto.so.1.1|
|Debian 9|/usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2|/usr/lib/x86_64-linux-gnu/libssl.so.1.0.2|
|Debian 10+|/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1|/usr/lib/x86_64-linux-gnu/libssl.so.1.1|

This symbolic link means that when the `libmi` library is loaded and the loader goes through the linked dependencies and finds `libssl.so.1.0.0` and `libcrypto.so.1.0.0` it will select the symbolic link in the PowerShell directory that the points to the proper proper that is specific for that distribution.

If any of these dependencies (as denoted by `ldd`) cannot be found then `libmi` will fail to load which also means `libpsrpclient` cannot be loaded by PowerShell causing the error

> This parameter set requires WSMan, and no supported WSMan client library was found. WSMan is either not installed or unavailable for this system.

From what I can gather the intent of this PR and what the current package symlinks target is to ensure that `libssl` and `libcrypto` libraries are pointing to an actual library specific for that distribution so `libmi` can be loaded and used by `libpsrpclient`. The problem with this approach is that `libmi` has been compiled against OpenSSL 1.0.x and not 1.1.x so anything in `libmi` that calls or references a symbol in `libssl` or `libcrypto` either may not exist (API) or the differ in how it is meant to be called (ABI) causing a failure and crashing the process with

> pwsh: symbol lookup error: /opt/microsoft/powershell/7/libmi.so: undefined symbol: SSL_library_init

This will occur for anything that tries to connect over a `https` endpoint over WSMan, like Exchange Online.

Ultimately there are a few options available:

* Continue as is and create a symlink pointing to OpenSSL 1.1.x on the distributions that default to that version
  * It is more likely WSMan will work on `http` out of the box but using `https` will crash the entire process
  * Gives a false indication that the WSMan libs work with OpenSSL 1.1.x and not the old and unsupported OpenSSL 1.0.x
  * The above doesn't matter as much because anything TLS related will just cause a crash but it is still an incorrect assumption
* Have the symlinks point to the OpenSSL 1.0.x path for all distributions
  * EL7 OpenSSL package is based on OpenSSL 1.0.2 so this is the out of box experience
  * EL8 has the `compat-openssl10` package which provides the OpenSSL 1.0.2 libs
  * Debian 9 has the `libssl1.0.2` package
  * Debian 10 doesn't have the `libssl1.0.2` package but you can technically just install it from the Debian 9 repo
  * Ubuntu is a bit harder but it [seems to be possible to install it](https://askubuntu.com/questions/989086/stuck-downgrading-libssl-on-ubuntu-16-04)
  * This ensures that TLS works with WSMan and people are explicitly aware that they are using an older and unsupported OpenSSL version
* Provide a new build of `libmi` with PowerShell that is compiled against OpenSSL 1.1.x
  * While the `omi` team doesn't support PowerShell I don't see why this stops PowerSell from using their "official" releases https://github.com/microsoft/omi/releases/tag/v1.6.6-0
  * They include a "universal" build for both OpenSSL 1.0.x and 1.1.x allowing it to work on both older and newer distributions
  * The PS team seems quite reluctant to using this but honestly I see this as a win win; PowerShell now uses OpenSSL 1.1.x which supports newer cipher suites and protocols and it works out of the box for more distributions
  *  It's not like the current distributed `libmi` in PowerShell is accepting any bugfixes so why not just use what the OMI team has already released which works with newer versions
  * There is a complication with `macOS` but that's a whole other story and I don't see why it should bring down Linux

I really hope the 3rd option is chosen, I just don't see any reason why PowerShell can't just use what is already provided by the OMI team instead of continuing to require OpenSSL 1.0.x to use WSMan over HTTPS.

Comment (iSazonov) [reply tools/packaging/packaging.strings.psd1:14]: @jborean93 Thanks for great post! But... I know that you known that MSFT does not support this scenario with libmi, they have officially announced that they will be removing libmi at all. (So DSC will be replaced by classes, and remoting will be transformed into a subsystem so that we can create modern alternatives.)

The PR is only for making universal Unix package and here is nothing about fixing any scenario.

Comment (TravisEz13) [reply tools/packaging/packaging.strings.psd1:14]: The goal here is not to improve our support for libmi.

File: tools/packaging/packaging.strings.psd1
Comment (TravisEz13) [reply tools/packaging/packaging.strings.psd1:52]: See https://github.com/PowerShell/PowerShell/pull/15109#discussion_r602769825

File: tools/packaging/packaging.strings.psd1
Comment (jborean93) [reply tools/packaging/packaging.strings.psd1:14]: So you are going to create completely new code that just replicates the existing, and known to be wrong, behaviour rather than fix it at the same time. Seems like a waste of effort but I’m tired of trying to advocate this be fixed so I’m giving up.

Comment (iSazonov) [reply tools/packaging/packaging.strings.psd1:14]: @jborean93 I agree with you that a working solution should have been made first 4 years ago, but this has not been done. I am at a loss and can only explain this by political wars. But this is no longer important, since the time has already passed. For now, we only have to insist on the implementation of a new remoting subsystem. This will untie the hands of both the community and MSFT itself.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by DHowett** [comment]
As a package maintainer, I'd like to lodge my dissent with making files (or symlinks, or permission changes) in postinst scripts. They can't be properly removed, the package manager doesn't know about them, and generally they circumvent the OS's mechanism for handling packaged data. :smile:

**Comment by iSazonov** [comment]
Perhaps @heaths has thoughts how improve this.

**Comment by jborean93** [comment]
As mentioned in https://github.com/PowerShell/PowerShell/pull/15109#discussion_r602666434 you could just remove the symlinks altogether as the current method is both pointing to the wrong SSL version and is trying to paper over a more fundamental problem. The only outlier is CentOS 7 which uses a slightly different library name but that could be done by patching the ELF fields for that distribution.

**Comment by TravisEz13** [comment]
@DHowett Please open an issue.  This objection needs to be reviewed.  As these symlinks are for a feature that eventually will no longer work, my proposal would be to just remove the links, but that breaks that feature for everyone.

Another alternative, to maintain compatibility, is to make a second package which this package has as an optional dependency on.

**Comment by heaths** [comment]
Is this question in response to tools/packaging/packaging.strings.psd1? As for build scripts, I see no problems with that and changes should be isolated (i.e. don't mess with the host outside of the build directory), but if in relation to the aforementioned file I really don't know. All I can say is that I don't see symlinking other packages' files by scripts that don't own them. Seems wrong to mess with another system like that - like in the old days before SFC on Windows when third-parties would replace Windows' bits.

**Comment by TravisEz13** [comment]
@heaths `libmi`, provided by another team, has a hard coded path to load the libraries from.  That team has decided not to support PowerShell anymore.  To keep things functioning (until it completely brakes), this is our work around.   symlinks would be our solution, if through the package manager or the script.  I'm honestly happy to pull the symlinks from the package all together.

I don't think this is the same as replacing windows bits.  We are symlinking other packages INTO our folder, not the other way around.

**Comment by jborean93** [comment]
> provided by another team, has a hard coded path to load the libraries from.

That's not strictly true, the "universal" Linux build they have provided you are linked to the following OpenSSL libs

* `libssl.so.1.0.0`
* `libcrypto.so.1.0.0`

You can see that in the following snippet, note I don't have OpenSSL 1.0.x installed and deleted the symlinks to show you what actual linked name so `not found` is expected.

```
[root@713b4cbcd41d /]# ldd /opt/microsoft/powershell/7/libmi.so
ldd: warning: you do not have execution permission for `/opt/microsoft/powershell/7/libmi.so'
        linux-vdso.so.1 (0x00007ffeb393a000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb157362000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fb15715e000)
        libpam.so.0 => /lib64/libpam.so.0 (0x00007fb156f4e000)
        libssl.so.1.0.0 => not found
        libcrypto.so.1.0.0 => not found
        libc.so.6 => /lib64/libc.so.6 (0x00007fb156b8b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb157582000)
        libaudit.so.1 => /lib64/libaudit.so.1 (0x00007fb156961000)
        libcap-ng.so.0 => /lib64/libcap-ng.so.0 (0x00007fb15675b000)
```

Typically these libraries are just picked up in the normal load path like `/usr/lib`, or the one specified by `LD_LIBRARY_PATH`. The `libmi.so` binary also has an RPATH set to `$ORIGIN` which allows it to also check the origin path (same dir it is placed in) which is what allows you to use the symlinks

```
[root@2cd102a3a066 /]# objdump -x /opt/microsoft/powershell/7/libmi.so | grep RPATH
  RPATH                $ORIGIN
```

For example on Ubuntu 16.04 if I delete the symlinks in the PowerShell package it still links just fine with the system provided libs because `/lib/x86_64-linux-gnu` is part of the load path to check:

```
root@9d42597878f5:/# ldd /opt/microsoft/powershell/7/libmi.so
        linux-vdso.so.1 =>  (0x00007ffc80db9000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0741cf0000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0741aec000)
        libpam.so.0 => /lib/x86_64-linux-gnu/libpam.so.0 (0x00007f07418de000)
        libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f0741676000)
        libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f0741231000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0740e67000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0741f0d000)
        libaudit.so.1 => /lib/x86_64-linux-gnu/libaudit.so.1 (0x00007f0740c40000)
```

In a normal situation you should just be picking up these libraries as they are installed, this actually works just fine if you have OpenSSL 1.0.x installed with only 2 caveats

* Enterprise Linux 7 (CentOS 7, RHEL 7, etc) uses `.10` as the suffix for the OpenSSL libs in `/usr/lib64` so they won't be picked up
  * This is one case where you probably need the symlinks or to patch the ELF headers to point to the distro specific path for that distribution only
  * Other distributions that still package OpenSSL 1.0.x use the `lib[ssl|crypto].so.1.0.0` file so they just work
* macOS does have a hardcoded path to `/usr/local/opt/openssl/lib/lib[ssl|crypto].1.0.0.dylib`
  * This is a different distribution package and IIRC it doesn't contain symlinks so I consider it a separate problem

The other issue is that the PowerShell package for newer distributions have a symlink to point `lib[ssl|crypto].1.0.0` to the OpenSSL 1.1.x paths which is completely wrong. This causes a seg fault in the process when `libmi` tries to use any OpenSSL APIs.

So ultimately what I think should happen is

* Patch the `libmi.so` library to change the suffix of libssl and libcrypto to `.10` for the RPM/Enterprise Linux package
* Do nothing, no symlinks on the other Linux distributions

I also think you should be consuming the newer releases from the [omi repo](https://github.com/microsoft/omi/releases) where that team has actually provided an official "universal" OpenSSL 1.1.x build which you can use on newer distributions that don't ship OpenSSL 1.0.x (read mostly all of them). It also means that patched `libmi.so` library is only shipped on EL7, everything else shouldn't require symlinks.

macOS is another problem and there's not much you can do there without recompling your own version of `omi` on OpenSSL 1.1.x which is understandable why you don't want to do.

**Comment by TravisEz13** [comment]
@jborean93 This issue is not to improve `libmi` behavior.  Please open a new issue for making libmi behavior better.


As for resolving this, I'll open a new issue for resolving creating the symlinks in the postscript.  Thinking about it, I think the best approach is separate packages with the symlink.

**Comment by jborean93** [comment]
I would love to resolve the libmi but I know the team doesn’t want to look into it. My posts are talking about the issues with how the current libmi is packaged. The package has symlinks which is not needed all the time. The package symlinks point to OpenSSL 1.1.x which causes set faults in PowerShell whenever you try to use any of the TLS components in libmi. How is that not related to the packaging components and thus this PR.

**Comment by daxian-dbw** [comment]
@jborean93 How about making the `PSWSMan` module the only source of `libmi` and `libpsrpclient` binaries for PowerShell remoting over WSMan on Linux and macOS? Namely, powershell will not ship those libs, but instead, tell users to leverage the `PSWSMan` module to enable WSMan remoting on Linux and macOS.

Please see the discussion topic https://github.com/PowerShell/PowerShell/discussions/15310, and let's discuss the feasibility there.

**Pull request merged**
